src/core/address_space.cpp

```diff
@@ -382,13 +393,16 @@ struct AddressSpace::Impl {
     void* Map(VAddr virtual_addr, PAddr phys_addr, size_t size, PosixPageProtection prot,
               int fd = -1) {
         m_free_regions.subtract({virtual_addr, virtual_addr + size});
-        const int handle = phys_addr != -1 ? (fd == -1 ? backing_fd : fd) : -1;
-        const off_t host_offset = phys_addr != -1 ? phys_addr : 0;
-        const int flag = phys_addr != -1 ? MAP_SHARED : (MAP_ANONYMOUS | MAP_PRIVATE);
-        void* ret = mmap(reinterpret_cast<void*>(virtual_addr), size, prot, MAP_FIXED | flag,
-                         handle, host_offset);
-        ASSERT_MSG(ret != MAP_FAILED, "mmap failed: {}", strerror(errno));
-        return ret;
+        if (phys_addr != -1) {
+            const int handle = fd == -1 ? backing_fd : fd;
+            void* ret = mmap(reinterpret_cast<void*>(virtual_addr), size, prot, MAP_FIXED | MAP_SHARED,
+                             handle, phys_addr);
+            ASSERT_MSG(ret != MAP_FAILED, "mmap failed: {}", strerror(errno));
+        } else {
+            int ret = mprotect(reinterpret_cast<void*>(virtual_addr), size, prot);
+            ASSERT_MSG(ret == 0, "mprotect failed: {}", strerror(errno));
```

Comment:
This and the below in `Unmap` will need a closer look to make sure the behavior is correct, I put it together for you quick to get things going but I'm not sure it's good enough to go upstream.

Reply:
Actually no, this part was @squidbus who has more experience with the memory, so I trust that it's correct. Feel free to test though

Reply:
That's... me though.